lib/cn/csched_sp3_work.c

```diff
@@ -362,20 +363,17 @@ sp3_work_node_idle(
     list_for_each_entry(le, head, le_link) {
         if (kvset_get_compc(le->le_kvset) >= idlec)
             break;
+
+        vgroups += kvset_get_vgroups(le->le_kvset);
         *mark = le;
         ++runlen;
     }
 
-    if (runlen < SP3_LLEN_RUNLEN_MIN)
+    if (runlen < SP3_LLEN_RUNLEN_MIN || runlen >= vgroups)
```

Comment:
Not sure I am understanding comparing runlen to vgroups here.

Reply:
A kvset that hasn't been k-compacted has no scatter (e.g., vgroups == 1).  If we have more kvsets than vgroups then this function will not create any work to compress these kvsets, it'll just leave them alone as they are already in pretty good shape.

Reply:
What is a case when runlen could be greater than vgroups? I can only come up with a scenario where runlen is equal to vgroups. I may have similar logic in cn-tree-simulator to protect against similar unnecessary work.

Reply:
If you have say four uncompressed kvsets with 1 vgroup each and kcompact them you'll end up with 1 kvset that contains four vgroups.  Repeat that three more times, and on the third repeat you'll have four kvsets each with four vgroups, hence runlen will be quite a bit less than then number of vgroups in that run of kvsets.
Now if instead you kv-compact some run of kvsets the resulting kvset will contain only 1 vgroup.